### PR TITLE
Inline mem::size_of & mem::align_of

### PR DESCRIPTION
**Pull request by tmiasko**
Opened for perf results.

cc @bjorn3
r? @ghost

### CODE REVIEW

File: library/core/src/alloc/layout.rs [hunk 1]
```diff
@@ -164,7 +164,8 @@ impl Layout {
     #[unstable(feature = "layout_for_ptr", issue = "69835")]
     pub unsafe fn for_value_raw<T: ?Sized>(t: *const T) -> Self {
         // SAFETY: we pass along the prerequisites of these functions to the caller
-        let (size, align) = unsafe { (mem::size_of_val_raw(t), mem::align_of_val_raw(t)) };
+        let size = size_of_val(t);
+        let align = align_of_val(t);
```
Comment (tmiasko) [inline library/core/src/alloc/layout.rs:168]: There is inconsistency in safety of size_of_val_raw function and intrinsic used to implement it.

**Comment by bugadani** [comment]
Just out of curiousity, looking at [webrenderer](https://perf.rust-lang.org/detailed-query.html?commit=d0efbf33ce34c88893bff82178722695e0031b9c&base_commit=fde692739576089729885b7f79aa2232cb9778c5&benchmark=webrender-wrench-opt&run_name=incr-unchanged) results, how can this PR change the number of executions for certain queries?

**Comment by tmiasko** [comment]
The recurring reduction by 4 corresponds to the situation where we avoided creating any monomorphiziations of given item: `size_align` would be one because it is removed, but I would expect that is also have happened for some of wrapper functions in mem module.

The `resolve_instance` reduction by 721 corresponds to the reduction in the number of calls terminators inside codegen items. When lowering pass was first enabled (it lowers `size_of` intrinsic call into a simple MIR statement) this number dropped by 328. This PR removes one layer of indirection when calling both size & align, which is roughly twice that.

**Comment by tmiasko** [comment]
This approach avoids the perf regression encountered before. I would be interested in potentially landing this, although I am not sure what are general opinions about using intrinsics directly like that.

File: library/alloc/src/collections/vec_deque/mod.rs [hunk 2]
```diff
@@ -58,7 +59,7 @@ mod tests;
 const INITIAL_CAPACITY: usize = 7; // 2^3 - 1
 const MINIMUM_CAPACITY: usize = 1; // 2 - 1
 
-const MAXIMUM_ZST_CAPACITY: usize = 1 << (core::mem::size_of::<usize>() * 8 - 1); // Largest possible power of two
+const MAXIMUM_ZST_CAPACITY: usize = 1 << (size_of::<usize>() * 8 - 1); // Largest possible power of two
```
Comment (the8472) [inline library/alloc/src/collections/vec_deque/mod.rs:62]: This could be updated to use `usize::BITS`

**Comment by oli-obk** [comment]
I don't feel comfortable with this extensive direct use of intrinsics. If we could turn intrinsics into regular function pointers, then we could do something like we did for `transmute` and re-export the intrinsic in `mem`. Since they have a different ABI, doing that without compatibility between `rust` and `rust-intrinsic` ABI would be a breaking change.

Maybe we could take a different approach and do it like https://github.com/rust-lang/rust/blob/ab5b9aecb9656a6d6e61fc7c14e4f1fe438dde56/library/core/src/ptr/mod.rs#L177-L185 and not have intrinsics for these at all (so they don't end up with a `rust-intrinsic` ABI), allowing us to take pointers to them, while still essentially treating them as intrinsics.

**Comment by bjorn3** [comment]
`drop_in_place` is codegened as a real function with real MIR. This real MIR just depends on the generic param unlike regular functions. Intrinsics are codegened inline at the caller site.

**Comment by oli-obk** [comment]
Hmm right, that is a bit different. So that scheme is too different I guess. While we could make mir building automatically generate the appropriate MIR statements at all call sites to specific lang items, that is something completely new that we haven't done so far.

**Comment by tmiasko** [comment]
If we were comfortable with using intrinsics directly this seems like nice win, but otherwise I wouldn't consider it to merit an extra compiler work, except for continued efforts towards enabling MIR inlining by default.

The area that does need compiler work are intrinsic wrappers in stdarch where overhead in the range of hundreds basic blocks is typical, in some cases going up to thousands of unnecessary basic blocks.

**Comment by oli-obk** [comment]
> The area that does need compiler work are intrinsic wrappers in stdarch where overhead in the range of hundreds basic blocks is typical, in some cases going up to thousands of unnecessary basic blocks.

Do you think the problem is that we build these blocks at all, or could we just run an early mir opt to inline all intrinsic wrappers?

**Comment by tmiasko** [comment]
> Do you think the problem is that we build these blocks at all, or could we just run an early mir opt to inline all intrinsic wrappers?

For the details regarding stdarch situation see https://github.com/rust-lang/stdarch/issues/248.

**Comment by oli-obk** [comment]
oof. that's a whole different situation imo.

**Comment by tmiasko** [comment]
Anyway, since there are reservation about using intrinsics directly I think that answers question about this proposal.

(I feel you might have misinterpreted my tangential comment about stdarch. I wasn't commenting about similarities, quite the contrary. I think the wrappers in stdarch would benefit for extra compiler work to make those functions trivial, unlike the wrappers here which I don't think would).

**Pull request closed without merging**